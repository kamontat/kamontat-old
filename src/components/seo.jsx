/**
 * SEO component that queries for data with
 *  Gatsby's useStaticQuery React hook
 *
 * See: https://www.gatsbyjs.org/docs/use-static-query/
 */

import React from "react";
import PropTypes from "prop-types";
import Helmet from "react-helmet";
import { useStaticQuery, graphql } from "gatsby";

function SEO({ description, lang, meta, title }) {
  const { site, twitter, github } = useStaticQuery(
    graphql`
      query {
        site {
          siteMetadata {
            title
            description
            app {
              version
            }
          }
        }
        twitter: contentfulSocial(name: { eq: "Twitter" }) {
          username
        }
        github: contentfulSocial(name: { eq: "Github" }) {
          username
          url
        }
      }
    `,
  );

  const metaDescription = description || site.siteMetadata.description;

  return (
    <Helmet
      htmlAttributes={{
        lang,
      }}
      title={title}
      defaultTitle={site.siteMetadata.title}
      titleTemplate={`%s | ${site.siteMetadata.title}`}
      meta={[
        {
          name: `version`,
          content: site.siteMetadata.app.version,
        },
        {
          name: `description`,
          content: metaDescription,
        },
        {
          property: `gh:account`,
          content: github.username,
          link: github.url,
        },
        {
          property: `og:title`,
          content: title,
        },
        {
          property: `og:description`,
          content: metaDescription,
        },
        {
          property: `og:type`,
          content: `website`,
        },
        {
          name: `twitter:card`,
          content: `summary`,
        },
        {
          name: `twitter:creator`,
          content: `@${twitter.username}`,
        },
        {
          name: `twitter:title`,
          content: title,
        },
        {
          name: `twitter:description`,
          content: metaDescription,
        },
      ].concat(meta)}
    />
  );
}

SEO.defaultProps = {
  lang: `en`,
  meta: [],
  description: ``,
};

SEO.propTypes = {
  description: PropTypes.string,
  lang: PropTypes.string,
  meta: PropTypes.arrayOf(PropTypes.object),
  title: PropTypes.string,
};

export default SEO;
