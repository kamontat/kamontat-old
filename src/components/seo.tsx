/**
 * SEO component that queries for data with
 *  Gatsby's useStaticQuery React hook
 *
 * See: https://www.gatsbyjs.org/docs/use-static-query/
 */

import React from "react";
import PropTypes from "prop-types";
import { Helmet } from "react-helmet";
import { useStaticQuery, graphql } from "gatsby";

import { DefaultProps } from "../typescript/properties";

type MetaType = JSX.IntrinsicElements["meta"];

interface SeoProps extends DefaultProps {
  title: string;
  description: string;
  lang: string;
  meta: MetaType[];
}

function SEO(props: SeoProps) {
  const { site, twitter, github } = useStaticQuery(
    graphql`
      query {
        site {
          siteMetadata {
            title
            description
            app {
              version
            }
          }
        }
        twitter: contentfulSocial(name: { eq: "Twitter" }) {
          username
        }
        github: contentfulSocial(name: { eq: "Github" }) {
          username
          url
        }
      }
    `,
  );

  const metaDescription = props.description || site.siteMetadata.description;

  return (
    <Helmet
      htmlAttributes={{
        lang: props.lang,
      }}
      title={props.title}
      defaultTitle={site.siteMetadata.title}
      titleTemplate={`%s | ${site.siteMetadata.title}`}
      meta={[
        {
          name: `version`,
          content: site.siteMetadata.app.version,
        } as MetaType,
        {
          name: `description`,
          content: metaDescription,
        } as MetaType,
        {
          property: `gh:account`,
          content: github.username,
          link: github.url,
        } as MetaType,
        {
          property: `og:title`,
          content: props.title,
        } as MetaType,
        {
          property: `og:description`,
          content: metaDescription,
        } as MetaType,
        {
          property: `og:type`,
          content: `website`,
        } as MetaType,
        {
          name: `twitter:card`,
          content: `summary`,
        } as MetaType,
        {
          name: `twitter:creator`,
          content: `@${twitter.username}`,
        } as MetaType,
        {
          name: `twitter:title`,
          content: props.title,
        } as MetaType,
        {
          name: `twitter:description`,
          content: metaDescription,
        } as MetaType,
      ].concat(props.meta)}
    />
  );
}

SEO.defaultProps = {
  lang: `en`,
  meta: [],
  description: ``,
  title: ``,
} as SeoProps;

SEO.propTypes = {
  description: PropTypes.string,
  lang: PropTypes.string,
  meta: PropTypes.arrayOf(PropTypes.object),
  title: PropTypes.string,
};

export default SEO;
