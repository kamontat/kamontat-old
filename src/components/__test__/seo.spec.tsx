import React from "react";
import { mount } from "enzyme";

import { PureSEO } from "../seo";
import Helmet from "react-helmet";

describe("Seo component", () => {
  const data = {
    title: "First title",
    description: "Description",
    lang: "en",
    meta: [],
    site: {
      siteMetadata: {
        title: "Second title",
        description: "Description",
        app: {
          name: "hello",
          version: "v1.0.0",
        },
        buildTime: "1586018684444",
      },
    },
    twitter: {
      username: "name",
    },
    github: {
      username: "ghname",
      url: "https://gh.com",
    },
  };

  it("render correctly", () => {
    const component = mount(<PureSEO data={data} />);

    const helmet = Helmet.peek();

    expect(helmet).toMatchSnapshot();

    component.unmount();
  });
});
