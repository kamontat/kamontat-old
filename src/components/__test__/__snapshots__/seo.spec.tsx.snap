// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Seo component render correctly: SEO component 1`] = `
Object {
  "baseTag": Array [],
  "bodyAttributes": Object {},
  "defer": true,
  "encode": true,
  "htmlAttributes": Object {
    "lang": "en",
  },
  "linkTags": Array [],
  "metaTags": Array [
    Object {
      "content": "hello",
      "name": "app:name",
    },
    Object {
      "content": "Description",
      "name": "app:description",
    },
    Object {
      "content": "v1.0.0",
      "name": "app:version",
    },
    Object {
      "content": "1586018684444",
      "name": "app:datetime",
    },
    Object {
      "content": "Invalid Date",
      "name": "app:datetime-formatted",
    },
    Object {
      "content": "Description",
      "name": "description",
    },
    Object {
      "content": "ghname",
      "link": "https://gh.com",
      "property": "gh:account",
    },
    Object {
      "content": "First title",
      "property": "og:title",
    },
    Object {
      "content": "Description",
      "property": "og:description",
    },
    Object {
      "content": "website",
      "property": "og:type",
    },
    Object {
      "content": "summary",
      "name": "twitter:card",
    },
    Object {
      "content": "@name",
      "name": "twitter:creator",
    },
    Object {
      "content": "First title",
      "name": "twitter:title",
    },
    Object {
      "content": "Description",
      "name": "twitter:description",
    },
  ],
  "noscriptTags": Array [],
  "onChangeClientState": [Function],
  "scriptTags": Array [],
  "styleTags": Array [],
  "title": "First title | Second title",
  "titleAttributes": Object {},
}
`;
